File: cppgit/command.cc
#include <cstdlib>
#include <ctime>
#include <filesystem>
#include <fstream>
#include <iterator>
#include <iostream>
#include <string>

#include "./author.hh"
#include "./blob.hh"
#include "./commit.hh"
#include "./database.hh"
#include "./entry.hh"
#include "./tree.hh"
#include "./workspace.hh"

namespace fs = std::filesystem;

namespace git {

void command(int argc, const char** argv) {
  using namespace std;

  string command(argv[1]);

  if (command == "init") {
    auto path = fs::absolute(argv[2]);
    auto git_path = path / ".git";

    try {
      for (auto dirname : {"objects", "refs"}) {
        fs::create_directories(git_path / dirname);
      }

      cerr << "Initialized empty git repository in " << git_path << endl;
    } catch (fs::filesystem_error& e) {
      cerr << "fatal: " << e.what() << endl;
    }
  } else if (command == "commit") {
    auto root_path = fs::absolute(".");
    auto git_path = root_path / ".git";
    auto db_path = git_path / "objects";

    auto ws = Workspace(root_path);
    auto database = Database(db_path);
    auto tree = Tree();
    for (auto& path : ws.ListFiles()) {
      auto content = ws.ReadFile(path);
      auto blob = Blob(content);
      database.Store(blob);
      content.data();
      tree.add_entry({fs::relative(path), blob.oid()});
      delete const_cast<void*>(content.data());
    }
    database.Store(tree);

    auto name = std::getenv("GIT_AUTHOR_NAME");
    auto email = std::getenv("GIT_AUTHOR_EMAIL");
    time_t now;
    time(&now);
    auto author = Author(name, email, now);

    cin >> std::noskipws;
    istream_iterator<char> it(cin);
    istream_iterator<char> end;
    string message(it, end);

    auto commit = Commit(tree.hex(), author, message);
    database.Store(commit);

    ofstream head;
    head.open(git_path / "HEAD");
    head << commit.hex();
    head.close();

    cout << "[(root-commit) " << commit.hex() << "] " << message << endl;
  } else {
    cerr << "git: '" << command << "' is not a git command." << endl;
    exit(1);
  }
}

}  // namespace git

File: gogit/git/command.go
package git

import "bufio"
import "fmt"
import "io/ioutil"
import "os"
import "path"
import "path/filepath"
import "time"

func Command(args []string) {
	command := args[1]
	if command == "init" {
		root_path, _ := filepath.Abs(os.Args[2])
		git_path := path.Join(root_path, ".git")
		for _, dir := range [3]string{"objects", "refs"} {
			new_path := path.Join(git_path, dir)
			merr := os.MkdirAll(new_path, os.ModePerm)
			check(merr)
		}
		fmt.Println("Initialized empty git repository in", git_path)
	} else if command == "commit" {
		root_path, _ := os.Getwd()
		git_path := path.Join(root_path, ".git")
		db_path := path.Join(git_path, "objects")

		ws := Workspace{root_path}
		database := new(Database).Init(db_path)
		entries := []Entry{}
		for _, path := range ws.ListFiles() {
			content := ws.ReadFile(path)
			blob := new(Blob).Init(content)
			fmt.Println(path)
			database.Store(blob)
			entries = append(entries, Entry{path, blob.Id()})
		}

		tree := new(Tree).Init(entries)
		database.Store(tree)
		fmt.Println(tree.Id())

		name := os.Getenv("GIT_AUTHOR_NAME")
		email := os.Getenv("GIT_AUTHOR_EMAIL")
		author := Author{name, email, time.Now()}

		reader := bufio.NewReader(os.Stdin)
		message, _ := reader.ReadString('\n')

		commit := new(Commit).Init(tree.Id(), author, message)
		database.Store(commit)

		hex := fmt.Sprintf("%x", commit.Id())
		err := ioutil.WriteFile(path.Join(git_path, "HEAD"), []byte(hex), 0644)
		check(err)

		fmt.Printf("[(root-commit) %x] %s\n", commit.Id(), message)
	} else {
		fmt.Printf("git: '%s' is not a git command.\n", command)
		panic(1)
	}
}
